PORT-7273 |  DORA Metrics (#1024)
port-labs/port-docs: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md

---
sidebar_position: 10
---

import PortTooltip from "/src/components/tooltip/tooltip.jsx";

> # Dora Metrics

Reviewer: ```suggestion
# DORA Metrics

```

> In this guide, we will create a self-service action in Port that executes a GitHub workflow to compute the dora metrics for a service.

Reviewer: ```suggestion
In this guide, we will create a self-service action in Port that executes a GitHub workflow to compute the DORA metrics for a service.

```

        "description": "This service's type",
        "type": "string",
        "enum": [
          "Backend",
          "Frontend",
          "Library"
        ],
        "enumColors": {
          "Backend": "purple",
          "Frontend": "pink",
          "Library": "green"
        },
        "icon": "DefaultProperty"
      },
      "lifecycle": {
        "title": "Lifecycle",
        "type": "string",
        "enum": [
          "Production",
          "Staging",
          "Development"
        ],
        "enumColors": {
          "Production": "green",
          "Staging": "yellow",
          "Development": "blue"
        },
        "icon": "DefaultProperty"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}
```
</details>

## Create Github workflow

Follow these steps to get started:

1. Create the following GitHub Action secrets:
    - `PORT_CLIENT_ID` - Port Client ID [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)
    - `PORT_CLIENT_SECRET` - Port Client Secret [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)

>     - `PATTOKEN` - GitHub PAT token. Ensure that Read access to actions and metadata permission is set. [learn more](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)

Reviewer: ```suggestion
    - `PATTOKEN` - GitHub PAT fine-grained token. Ensure that read-only access to actions and metadata permission is set. [learn more](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token)

```

Reviewer: I understand from the mentioned permissions that the token needs to be a fine-grained one. If so, the user needs more information such as - to which repositories should he give the token access

2. Create an action in the [self-service page](https://app.getport.io/self-serve) on the `Service` blueprint with the following JSON definitions:

<details>

  <summary><b>Port Action: Dora Metrics (click to expand)</b></summary>
   :::tip
- `<GITHUB-ORG>` - your GitHub organization or user name.
- `<GITHUB-REPO-NAME>` - your GitHub repository name.
:::

```json showLineNumbers
{
  "identifier": "dora_metrics",

>   "title": "Dora Metrics",

Reviewer: ```suggestion
  "title": "DORA Metrics",

```

  "icon": "Github",
  "userInputs": {
    "properties": {
      "timeframe": {
        "icon": "Github",
        "title": "Timeframe",
        "description": "Time frame in weeks to calculate metrics on",
        "type": "number",
        "default": 4,
        "minimum": 1
      },
      "workflow": {
        "title": "workflow",
        "description": "The name of the workflows to process. Multiple workflows can be separated by a comma (,) .",
        "icon": "Github",
        "type": "string"
      },
      "repository": {
        "title": "Repository",
        "format": "url",
        "type": "string",
        "icon": "Github",
        "default": {
          "jqQuery": ".entity.properties.url"
        }
      }
    },
    "required": [
      "timeframe",
      "workflow"
    ],
    "order": [
      "timeframe",
      "workflow"
    ]
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<GITHUB-ORG>",
    "repo": "<GITHUB-REPO-NAME>",
    "workflow": "dora-metrics.yaml",
    "omitUserInputs": false,
    "omitPayload": false,
    "reportWorkflowStatus": true
  },
  "trigger": "DAY-2",

>   "description": "Estimate dora metrics for a service",

Reviewer: ```suggestion
  "description": "Calculate DORA metrics for a service",

```

  "requiredApproval": false
}
```

</details>

3. In your Github repository, create a workflow file under `.github/workflows/dora-metrics.yml` with the following content:

<details>

<summary><b>GitHub workflow: Ingest DORA Metrics (click to expand)</b></summary>

```yaml showLineNumbers title="dora-metrics.yml"
name: Ingest DORA Metrics

on:
  workflow_dispatch:
    inputs:
      repository:
        description: 'Comma-separated list of repositories to analyze (eg. https://github.com/port-labs/self-service-actions)'
        required: true
      timeframe:
        description: 'Time frame within which metrics should be computed. e.g. 4 for Last 4 weeks'
        required: true
      workflow:
        description: 'The name of the workflows to process. Multiple workflows can be separated by a comma (,).'
        required: true
      port_payload:
        required: true
        description: 'Port's payload, including details for who triggered the action and
          general context (blueprint, run id, etc...)'
        type: string
      
jobs:
  compute-dora-metrics:
    runs-on: ubuntu-latest
    steps:
    
      - name: Checkout code

>         uses: actions/checkout@v2

Reviewer: ```suggestion
        uses: actions/checkout@v4

```

        
      - name: Transform Workflow Inputs
        run: |
          days=$(( ${{ github.event.inputs.timeframe }} * 7 ))
          repository_path=$(echo "${{ github.event.inputs.repository }}" | awk -F'com/' '{print $NF}')
          # repo_name="${repository_path##*/}"
          cleaned_name=$(echo "${repository_path##*/}" | tr -c '[:alnum:]' ' ')
          title=$(echo "$cleaned_name" | awk '{for(i=1;i<=NF;i++) $i=toupper(substr($i,1,1)) tolower(substr($i,2));}1')

          # export to github enviroment vars
          echo "TIMEFRAME_IN_DAYS=$days" >> $GITHUB_ENV
          echo "REPOSITORY=$repository_path" >> $GITHUB_ENV
          echo "TITLE=$title" >> $GITHUB_ENV
        shell: bash

      - name: Set up Python

>         uses: actions/setup-python@v2

Reviewer: ```suggestion
        uses: actions/setup-python@v5

```

      - name: Report Success In Computing PR Metrics
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Successfully Computed Pull Request metric for the service ✅"
          
      - name: Log Before Running Deployment Frequency
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Computing deployment frequency metric ..."
          
      - name: Deployment Frequency
        id: deployment_fequency
        env:
          WORKFLOWS: ${{ inputs.workflow }}
          GITHUB_TOKEN: ${{ secrets.PATTOKEN }} 
          REPOSITORY: ${{ env.REPOSITORY }}

>         run: python dora/deploymentfrequency.py

Reviewer: ```suggestion
        run: python dora/deployment_frequency.py

```

          
      - name: Report Success In Computing Deployment Frequency
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Successfully computed deployment frequency ✅"
          
      - name: Log Before Running Lead Time for Changes
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Computing lead time for changes metric ..."
          
      - name: Lead Time For Changes
        id: lead_time_for_changes
        env:
          WORKFLOWS: ${{ inputs.workflow }}
          GITHUB_TOKEN: ${{ secrets.PATTOKEN }} 
          REPOSITORY: ${{ env.REPOSITORY }}

>         run: python dora/leadtimeforchanges.py

Reviewer: ```suggestion
        run: python dora/lead_time_for_changes.py

```

        if: failure()
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Failed to upsert entity to port ❌"
          
      - name: Report Successful Upserting of Entity
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Entity upserting was successful ✅"
```

</details>
<br />

4. Create a text file (`requirements.txt`) in a folder name `dora` to host the required dependencies for running the action.
<details>
  <summary><b>Requirements</b></summary>

```text showLineNumbers title="requirements.txt"
PyGithub==2.3.0
requests==2.31.0
pytz==2024.1
```

</details>

> 5. Create the following python scripts (`calculate_pr_metrics.py`, `deploymentfrequency.py` and `leadtimeforchanges.py` ) in a folder named `dora` at the root of your GitHub repository:

Reviewer: ```suggestion
5. Create the following python scripts (`calculate_pr_metrics.py`, `deployment_frequency.py` and `lead_time_for_changes.py` ) in a folder named `dora` at the root of your GitHub repository:

```

            else 0,
            "average_commits_per_pr": round(aggregated["total_commits"]
            / aggregated["prs_opened"],2)
            if aggregated["prs_opened"]
            else 0,
            "average_loc_changed_per_pr": round(aggregated["total_loc_changed"]
            / aggregated["prs_opened"],2)
            if aggregated["prs_opened"]
            else 0,
            "average_prs_reviewed_per_week": round(average_prs_reviewed_per_week,2),
        }

        return metrics

    def timedelta_to_decimal_hours(self, td):
        return round(td.total_seconds() / 3600, 2)

def main():
    repo_name = os.getenv("REPOSITORY")
    time_frame = os.getenv("TIMEFRAME_IN_DAYS")
    print("Repository Name:", repo_name)
    print("TimeFrame (in weeks):", time_frame)

    repo_metrics = RepositoryMetrics(repo_name, time_frame)
    metrics = repo_metrics.calculate_pr_metrics()

    metrics_json = json.dumps(metrics, default=str)
    with open(os.getenv("GITHUB_ENV"), "a") as github_env:
        github_env.write(f"metrics={metrics_json}\n")

if __name__ == "__main__":
    main()
```
</details>

<details>
  <summary><b>Deployment Frequency</b></summary>

> ```python showLineNumbers title="deploymentfrequency.py"

Reviewer: ```suggestion
```python showLineNumbers title="deployment_frequency.py"

```

            "rating": rating,
            "number_of_unique_deployment_days": len(unique_dates),
            "total_deployments": len(workflow_runs_list)
        }

        print(f"Owner/Repo: {self.owner}/{self.repo_name}")
        print(f"Workflows: {', '.join(self.workflows)}")
        print(f"Branch: {self.branch}")
        print(f"Number of days: {self.number_of_days}")
        print(f"Deployment frequency over the last {self.number_of_days} days is {deployments_per_day} per day")
        print(f"Rating: {rating} ({color})")
        return json.dumps(results, default=str)

if __name__ == "__main__":
    owner_repo = os.getenv('REPOSITORY')
    token = os.getenv('GITHUB_TOKEN')
    workflows = os.getenv('WORKFLOWS')
    branch = 'main'
    time_frame = int(os.getenv('TIMEFRAME_IN_DAYS'))
    number_of_days = 30 if not time_frame else time_frame
    
    df = DeploymentFrequency(owner_repo,workflows, branch, number_of_days, pat_token=token)
    report = df.report()
    
    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
        github_env.write(f"deployment_frequency_report={report}\n")
```
</details>

<details>
  <summary><b>Lead Time For Changes</b></summary>

> ```python showLineNumbers title="leadtimeforchanges.py"

Reviewer: ```suggestion
```python showLineNumbers title="lead_time_for_changes.py"

```

    lead_time_for_changes_in_hours = pr_average + workflow_average
    print(f"PR average time duration: {pr_average} hours")
    print(f"Workflow average time duration: {workflow_average} hours")
    print(f"Lead time for changes in hours: {lead_time_for_changes_in_hours}")

    report = {
            "pr_average_time_duration" : round(pr_average,2),
            "workflow_average_time_duration" : round(workflow_average,2),
            "lead_time_for_changes_in_hours": round(lead_time_for_changes_in_hours,2)
    }
    rating = calculate_rating(lead_time_for_changes_in_hours)
    report.update(rating)
    
    return json.dumps(report, default=str)
    
if __name__ == "__main__":
    owner_repo = os.getenv('REPOSITORY')
    token = os.getenv('GITHUB_TOKEN')
    workflows = os.getenv('WORKFLOWS')
    branch = 'main'
    time_frame = int(os.getenv('TIMEFRAME_IN_DAYS'))
    number_of_days = 30 if not time_frame else time_frame
    
    report = main(owner_repo, workflows, branch, number_of_days)
    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
        github_env.write(f"lead_time_for_changes_report={report}\n")
```
</details>

> 6. Trigger the action from the [self-service](https://app.getport.io/self-serve) page of your Port application.

Reviewer: Please add in the doc an example of inputs to trigger the action with, as I believe users will have a hard time understanding what are the correct inputs for the `workflow` and `repository` inputs in the action form

        },
        "icon": "DefaultProperty"
      }
    },
    "required": []
  },
  "mirrorProperties": {},
  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}
```
</details>

## Create Github workflow

Follow these steps to get started:

1. Create the following GitHub Action secrets:
    - `PORT_CLIENT_ID` - Port Client ID [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)
    - `PORT_CLIENT_SECRET` - Port Client Secret [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)
    - `PATTOKEN` - GitHub PAT token. Ensure that Read access to actions and metadata permission is set. [learn more](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)
2. Create an action in the [self-service page](https://app.getport.io/self-serve) on the `Service` blueprint with the following JSON definitions:

<details>

  <summary><b>Port Action: Dora Metrics (click to expand)</b></summary>
   :::tip
- `<GITHUB-ORG>` - your GitHub organization or user name.
- `<GITHUB-REPO-NAME>` - your GitHub repository name.
:::

```json showLineNumbers

> [

Reviewer 2: Remove the square bracket and the top and end of this JSON

        "format": "url",
        "type": "string",
        "icon": "Github",
        "default": {
          "jqQuery": ".entity.properties.url"
        }
      }
    },
    "required": [
      "timeframe",
      "workflow"
    ],
    "order": [
      "timeframe",
      "workflow"
    ]
  },
  "invocationMethod": {
    "type": "GITHUB",
    "org": "<GITHUB-ORG>",
    "repo": "<GITHUB-REPO-NAME>",
    "workflow": "dora-metrics.yaml",
    "omitUserInputs": false,
    "omitPayload": false,
    "reportWorkflowStatus": true
  },
  "trigger": "DAY-2",
  "description": "Estimate dora metrics for a service",
  "requiredApproval": false
}
]
```

</details>

3. In your Github repository, create a workflow file under `.github/workflows/dora-metrics.yml` with the following content:

<details>

<summary><b>GitHub workflow: Ingest DORA Metrics (click to expand)</b></summary>

```yaml showLineNumbers title="dora-metrics.yml"
name: Ingest DORA Metrics

on:
  workflow_dispatch:
    inputs:
      repository:
        description: 'Comma-separated list of repositories to analyze (eg. https://github.com/port-labs/self-service-actions)'
        required: true
      timeframe:

>         description: 'Last X weeks'

Reviewer 2: Give it a better description

        required: true
      workflow:

>         description: The name of the workflows to process. Multiple workflows can be separated by a comma (,).

Reviewer 2: Why are some `description` in single quote and some don't have quote at all?

        required: true
      port_payload:
        required: true
        description: Port's payload, including details for who triggered the action and
          general context (blueprint, run id, etc...)
        type: string
      
jobs:
  compute-dora-metrics:
    runs-on: ubuntu-latest
    steps:
    
      - name: Checkout code
        uses: actions/checkout@v2
        
      - name: Transform Workflow Inputs
        run: |
          days=$(( ${{ github.event.inputs.timeframe }} * 7 ))
          repository_path=$(echo "${{ github.event.inputs.repository }}" | awk -F'com/' '{print $NF}')
          # repo_name="${repository_path##*/}"
          cleaned_name=$(echo "${repository_path##*/}" | tr -c '[:alnum:]' ' ')
          title=$(echo "$cleaned_name" | awk '{for(i=1;i<=NF;i++) $i=toupper(substr($i,1,1)) tolower(substr($i,2));}1')

          # export to github enviroment vars
          echo "TIMEFRAME_IN_DAYS=$days" >> $GITHUB_ENV
          echo "REPOSITORY=$repository_path" >> $GITHUB_ENV
          echo "TITLE=$title" >> $GITHUB_ENV
        shell: bash

      - name: Set up Python
        uses: actions/setup-python@v2
        with:
          python-version: '3.x'

>       - name: Report Failure In Settting Up Dependencies

Reviewer 2: The name of the step can be `Notify intent to install dependencies`

        if: failure()
        uses: port-labs/port-github-action@v1
        with:
          clientId: ${{ secrets.PORT_CLIENT_ID }}
          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
          baseUrl: https://api.getport.io
          operation: PATCH_RUN
          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
          logMessage: "Installing required packages for the action..."
          
      - name: Install Python dependencies
        run: |
          python -m pip install --upgrade pip
          pip install -r dora/requirements.txt

>       - name: Report Failure In Computing PR Metrics

Reviewer 2: The name of the step should be `Report Failure In Settting Up Dependencies`

          logMessage: "Entity upserting was successful ✅"
```

</details>
<br />

4. Create a text file (`requirements.txt`) in a folder name `dora` to host the required dependencies for running the action.
<details>
  <summary><b>Requirements</b></summary>

```text showLineNumbers title="requirements.txt"
PyGithub==2.3.0
requests==2.31.0
pytz==2024.1
```

</details>

5. Create the following python scripts (`calculate_pr_metrics.py`, `deploymentfrequency.py` and `leadtimeforchanges.py` ) in a folder named `dora` at the root of your GitHub repository:

<details>
  <summary><b>Calculate PR Metrics</b></summary>

```python showLineNumbers title="calculate_pr_metrics.py"
import os
from github import Github
from datetime import datetime, timedelta, timezone
from concurrent.futures import ThreadPoolExecutor, as_completed
import json

class RepositoryMetrics:
    def __init__(self, repo_name, time_frame):
        self.github_client = Github(os.getenv("GITHUB_TOKEN"))
        self.repo_name = repo_name
        self.time_frame = int(time_frame)
        self.start_date = datetime.utcnow().replace(tzinfo=timezone.utc) - timedelta(
            days=self.time_frame
        )

>         self.repo = self.github_client.get_repo(f"{self.repo_name}")

Reviewer 2: why can't you do this? 
self.github_client.get_repo(self.repo_name)

            / aggregated["prs_opened"],2)
            if aggregated["prs_opened"]
            else 0,
            "average_loc_changed_per_pr": round(aggregated["total_loc_changed"]
            / aggregated["prs_opened"],2)
            if aggregated["prs_opened"]
            else 0,
            "average_prs_reviewed_per_week": round(average_prs_reviewed_per_week,2),
        }

        return metrics

    def timedelta_to_decimal_hours(self, td):
        return round(td.total_seconds() / 3600, 2)

def main():
    repo_name = os.getenv("REPOSITORY")
    time_frame = os.getenv("TIMEFRAME_IN_DAYS")
    print("Repository Name:", repo_name)
    print("TimeFrame (in weeks):", time_frame)

    repo_metrics = RepositoryMetrics(repo_name, time_frame)
    metrics = repo_metrics.calculate_pr_metrics()

    metrics_json = json.dumps(metrics, default=str)
    with open(os.getenv("GITHUB_ENV"), "a") as github_env:
        github_env.write(f"metrics={metrics_json}\n")

if __name__ == "__main__":
    main()
```
</details>

<details>
  <summary><b>Deployment Frequency</b></summary>

```python showLineNumbers title="deploymentfrequency.py"
from github import Github
import datetime

> import pytz  # Make sure to install pytz if you haven't: pip install pytz

Reviewer 2: If pytz is a required dependency then add it to the requirements.txt

Author: Its in the dependencies

    lead_time_for_changes_in_hours = pr_average + workflow_average
    print(f"PR average time duration: {pr_average} hours")
    print(f"Workflow average time duration: {workflow_average} hours")
    print(f"Lead time for changes in hours: {lead_time_for_changes_in_hours}")

    report = {
            "pr_average_time_duration" : round(pr_average,2),
            "workflow_average_time_duration" : round(workflow_average,2),
            "lead_time_for_changes_in_hours": round(lead_time_for_changes_in_hours,2)
    }
    rating = calculate_rating(lead_time_for_changes_in_hours)
    report.update(rating)
    
    return json.dumps(report, default=str)
    
if __name__ == "__main__":
    owner_repo = os.getenv('REPOSITORY')
    token = os.getenv('GITHUB_TOKEN')
    workflows = os.getenv('WORKFLOWS')
    branch = 'main'
    time_frame = int(os.getenv('TIMEFRAME_IN_DAYS'))
    number_of_days = 30 if not time_frame else time_frame
    
    report = main(owner_repo, workflows, branch, number_of_days)
    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
        github_env.write(f"lead_time_for_changes_report={report}\n")
```
</details>

> 6. Trigger the actions from the [self-service](https://app.getport.io/self-serve) page of your Port application.

Reviewer 2: Trigger the action not `actions`
